blockstore/apps/bundles/models.py

```diff
@@ -0,0 +1,186 @@
+#    )
+
+    # For pretty URLs?
+    slug = models.SlugField(allow_unicode=True)
+
+    title = models.CharField(max_length=MAX_CHAR_FIELD_LENGTH, db_index=True)
+    description = models.TextField(max_length=10_000)
+
+    # We use a BinaryField(16) because Django's built-in UUID field is double
+    # the size it needs to be (32 vs. 16 bytes) on MySQL. We use url-safe base64
+    # encoding when passing around serialized references.
+    uuid = models.UUIDField(default=uuid.uuid4, unique=True)
+
+    def __str__(self):
+        return "Bundle {} - {}".format(self.uuid, self.slug)
+
+class BundleVersion(models.Model):
+    """
+    You should never create a BundleVersion manually -- use Bundle to create a
+    version through it's BundleDataStore and let this be created through
+    signal handling.
+
+    Should this be moved to a lower layer entirely?
+    """
+    id = models.BigAutoField(primary_key=True)
+    bundle = models.ForeignKey(
+        Bundle, related_name="versions", related_query_name="version"
+    )
+    version_num = models.PositiveIntegerField()
+
+    snapshot_hash = models.BinaryField(max_length=20, db_index=True)
+
+    # version_hash = Unique representation of state that is not just numbered?
+    # compatibility_hash = Let content determine a hash so that two versions can
+    #                      be compared for compatibility (ORA2 does this). A
+    #                      proxy for breaking change or major/minor versions.
+
+    change_description = models.TextField(max_length=1_000, blank=True)
+
+    @staticmethod
+    @receiver(snapshot_created)
+    def listen_for_snapshot_creation(**kwargs):
+        bundle_uuid = kwargs['bundle_uuid']
+        snapshot_hash = kwargs['snapshot_hash']
+
+        bundle = Bundle.objects.get(uuid=bundle_uuid)
+        versions = list(bundle.versions.order_by('-version_num')[:1])
+        next_version_num = versions[0].version_num + 1 if versions else 1
+
+        bundle.versions.create(
+            version_num=next_version_num,
+            snapshot_hash=snapshot_hash.digest(),
+        )
+
+
+    def __str__(self):
+        return str(BundleVersionKey(self.bundle.uuid, self.version_num))
+
+    class Meta:
+        unique_together = (
+            ("bundle", "version_num"),
+        )
+
+
+class BundleLink(models.Model):
+    """
+    A lightweight representation of the Link between two Bundles.
+
+    The most common lookup use case for this will be to notify or report on all
+    the Bundles that link *to* a particular Bundle, for notification and
+    reporting purposes. We track things at the Bundle level of granularity
+    because doing so at the BundleVersion level would take up much more space
+    with limited benefits. All precise version dependencies are captured at
+    the BundleData layer, where it can be represented more cheaply.
+    """
+    id = models.BigAutoField(primary_key=True)
+
+    # Would love to have better names here
```

Comment:
`lender` and `borrower` are probably the least ambiguous though, which is a very nice property when it comes to links. Maybe `source` and `consumer` if you just want some suggestions on different words though.